CIP-???? | Post Quantum signatures and native wallets (#441)
cardano-foundation/CIPs: CIP-????/README.md

Support post-quantum signatures (Dilithium+CRYSTALS, FALCON) as recommended by NIST standardization effort in Plutus Core;
allow validation of such signatures as builtins.
NIST plans to finalize standardization in 2023/2024 time frame, with US executive branch recommending
all projects to plan for post-quantum resistance by mid-2023.
We should prepare Cardano Blockchain for migration.

## Abstract

Provides a way of verifying post-quantum signatures as recommended by NIST in Plutus Core.
New builtins will allow use of these signatures within smart contracts. These builtins work over
`BuiltinByteString`s.

## Motivation

NIST estimates that pre-quantum cryptographic algorithms may be no longer safe
by 2035. However, the smart contracts have long development and lifetime,
so migration to post-quantum cryptography algorithms should start several
years before they would become insecure.
This summer NIST has recommended three PQC signature algorithms
for standardization, publishing their reference implementations,
and finishing initial comment period.

Supporting PQC signatures will bring the following benefits:

* smart contracts may use future standard algorithms today;

> * early support for PQC will simplify future PQC migration, if needed;

Reviewer: it's unclear to me what the PCQ migration is and how this enables it? I would like to see this elaborated somewhere. It seems that you have an idea of what it is and how we would do it, but it's not explained anywhere.

* future PQC migration would be smooth and require less changes;
* support for PQC may increase industry acceptance for long running projects;
* users that want their wallets to survive over ten years may
  use smart contracts verifying multiple algorithms today;
* Cardano will become first blockchain to verify PQC standard signatures,
  thus getting first-mover advantage.

## Use cases

> Adding multiple PQC signature algorithms will allow the following:

Reviewer: ```suggestion
Adding multiple PQC signature algorithms to Plutus Core will allow the following:
```

* opt-in migration of wallets and smart contracts to safer cryptographic protocol

> * pay-as-you-go for longer signatures required for post-quantum resistance

Reviewer: again, not sure what this means

* opt-in resistance to _cryptocalypse_ - surprise breakage of a single major encryption algorithm -
  by using multiple signature algorithms for redundancy.
* compliance with US government regulations past July 2023, when all agencies
  are expected to plan post-quantum migration.

## Specification

> [`DSIGN`]() class instance should be provided for the post-quantum signature algorithms.

Reviewer: ```suggestion
### `cardano-base`

[`DSIGN`]() class instance should be provided for the post-quantum signature algorithms.
```

### Plutus API

Two new builtin functions would be provided:

* A verification function for both pre-quantum and post-quantum resistant
  signature schemes:
     - CRYSTALS(DILITHIUM) signatures;
     - FALCON signatures;
     - SPHINCS signatures; and
     - SEQUOA

> * Extensible interface for signature verification functions.

Reviewer: Same comment as in the CPS: it's not clear why this is entangled with the PQC stuff.

* A long-term signature verification API useful for smart contract scripts,
  so that smart contracts do not need to change when new cryptographic protocol
  is added.
* Extensible interface for hash verification functions,
  so that smart contracts do not need to change when hashing algorithm
  is upgraded.

These would be based on NIST reference implementations,
as maintained by [`PQClean`](https://github.com/PQClean/PQClean) library.
This is a reference implementation of the three kinds of signature schemes in C. This 
implementation would be called from Haskell using direct bindings to C. These
bindings would be defined in `cardano-base`, using its existing `DSIGN`
interface, with new builtins in Plutus Core on the basis of the `DSIGN`
interface for both schemes.

The builtins would be costed as follows: each signature verification has
linear cost proportional to the number of message bits, and square with respect to key length.
_Constants will be proportional to the ratio of time as benchmarked against

> Ed25519._

Reviewer: I think you are suggesting a very complex cost model, namely something that branches on the first argument and then is costed differently (at least wrt constants) depending on that. We could probably add such a thing, but it seems complicated.

Since we expect adding more cryptographic signature algorithms as they evolve,
we should add datatype for selecting cryptographic algorithm:

```
data SignAlgo = Ed25519
              | Blake2
              | EcDSA
              | SECP256k1
              | Schnorr
              | Falcon
              | Sphincs
              | Dilithium
              | Sequoa
```

> To facilitate extensibility, the name of algorithm should be rendered as case-insensitive text in JavaScript

Reviewer: I'm confused about what this has to do with the rest of the proposal?

| Dilithium2              |               1312 |              2420 |                                 0.5 | [Selected](https://csrc.nist.gov/projects/post-quantum-cryptography/selected-algorithms-2022) |
| Sphincs+                |                 32 |              8080^[https://sofiaceli.com/2022/07/05/pq-signatures.html#:~:text=A%20table%20of-,comparison%20of%20sizes,-%3A] |                                 2.8^[Benchmark was for Sphins\*128s] | [Selected](https://csrc.nist.gov/projects/post-quantum-cryptography/selected-algorithms-2022) |
| sequoa863pq128          |                108 |               216 |                                    ? | Submitted^[Communication on ebats@cryp.to mailing list.] |

  : Reference for signature, key sizes, and time cost multipliers [as suggested here](https://blog.cloudflare.com/nist-post-quantum-surprise/). In the order of preference, with comparison against current default signature algorithm (Ed25519).

The following function should then be used to verify arbitrary binary signed
with permitted signature scheme:

```

> verifySignWithAlgo :: SignAlgo

Reviewer: I think this is the heart of the "extensibility" proposal. The obvious question is "why not just write this function in user-space, having it delegate to the appropriate single-algorithm builtin?"

Here's my guess at the scenario you're trying to support where that wouldn't be enough (please actually write it down yourself though, I'd prefer not to guess!):

1. Dapp creator writes script S using `verifySignWithAlgo`
2. S can dynamically (?) choose which signature algorithm it is using, maybe by doing something with a value in the datum (?)
3. We update the Plutus ledger language being used by S such that `verifySignWithAlgo` has additional algorithms (I'm actually not sure we can do this safely!)
4. S can be used as-is, only requiring a change to e.g. its datum to get it to use the new algorithm.

This has several tricky steps, notably 2 and 3. It's hard for me to say whether or not we could do that since I'm not sure what you're trying to do, but I think we might not be able to do those (esp 3) without e.g. making some scripts succeed that previously would have failed at runtime, which we shouldn't really do.

                  -> BuiltinByteString -- ^ public key
                  -> BuiltinByteString -- ^ message or hash
                  -> BuiltinByteString -- ^ signature
                  -> BuiltinBool
```

This function takes parameters of a specific part of the signature scheme, even
though they are all encoded as `BuiltinByteString`s. In order, for both
functions, these are:

1. A verification key (length depends on the algorithm chosen);
2. An input to verify (either the message itself, and a hash);
3. A signature.

The two different schemes handle deserialization internally: specifically, there
is a distinction made between 'external' representations, which are expected as
arguments, and 'internal' representations, used only by the implementations
themselves. This creates different expecations for each argument for both of

> these schemes; we describe these below.

Reviewer: where? I couldn't find it. Generally I'm confused about this

For the [all signature
schemes](),

> the keys are required to be given in the same format as NIST test vectors.

Reviewer: Please be specific or link to the specific description.

The builtin operations will error with a descriptive message if given inputs
that don't correspond to the constraints above, return `False` if the signature
fails to verify the input given the key, and `True` otherwise.

### Post-quantum wallets

> We should permit using all chosen signature algorithms to lock transactions.

Reviewer: :rotating_light: :rotating_light: :rotating_light: 

I assume you mean "to lock UTXOs".

This is A HUGE CHANGE and not a Plutus one. This is now veering into changing all the cryptography in Cardano as a whole. I _highly_ recommend making a separate CIP for that, as I think it will be extremely difficult and contentious and needs a lot of support.

cardano.{walletName}.hashAlgo
```

There should be additional extension registered if [CIP on wallet extensibility](https://github.com/cardano-foundation/CIPs/pull/462).

Wallet providers may choose to automatically generate ultra-secure wallets for multiple quantum signature
algorithms from a single seed. In such case, they are encouraged to provide a visible
hint of what is the current preferred.

### Use of standard hash algorithm

Note that use of hash would require use of quantum-resistant hash algorithm.
[SHA2, SHA3, BLAKE2 are considered quantum-safe](https://cryptobook.nakov.com/quantum-safe-cryptography#quantum-safe-and-quantum-broken-crypto-algorithms),
in particular standard 256-bit hash used in Cardano [should be safe for a long time](https://cryptobook.nakov.com/quantum-safe-cryptography#quantum-safe-and-quantum-broken-crypto-algorithms).
This should be documented in the Plutus documentation of `verifySignature`.

However, in the interest of long-term API evolution we propose to add a flexible hashing interface:

```
data SignAlgo = SHA2_256
              | SHA3_256
              | Blake2

hashWithAlgo :: HashAlgo -> BuiltinByteString -> BuiltinByteString
```

> This will also allow for gradual _opt-in_ migration of scripts as the blockchain evolves.

Reviewer: Again, this is about extensibilty, and I still think it should be clearly separated out and argued for.

```
data LTSignature = LongTermSignature {
                     ltsHashAlgo :: HashAlgo
                   , ltsSignAlgo :: SignAlgo
                   , ltsSignature :: BuiltinByteString
                   }

verifyLongTermSignature :: BuiltinByteString                       -- ^ public key
                        -> LongTermSignature                       -- ^ message or hash
                        -> BuiltinByteString                       -- ^ signature
                        -> Bool  -- is hash valid

isLongTermSignatureSecure :: LongTermSignature -> Bool
```

Supporting evolution of hash algorithms is much simpler:

```
-- | Long term hash
data LongTermHash = LongTermHash {
                      hashAlgo :: HashAlgo
                    , hash     :: BuiltingByteString
                    }

mostSecureHashAlgo :: HashAlgo

mostSecureHash :: BuiltinByteString -> LongTermHash
mostSecureHash = hashWithAlgo mostSecureHash
```

This API needs additional support from the off-chain applications in order to:

* provide correct public key for the given algorithm
* help users to transfer their euTxOs to new signature algorithms when old algorithms are obsoleted.

> Note that `LongTermHash` and `LongTermSignature` can be implemented as standard library functions.

Reviewer: So this is just a proposal to add some stuff to the `plutus` standard library? That doesn't need a CIP! I suggest removing it to keep things slim and opening an issue in `plutus`.

> ## Rationale

Reviewer: I would like to see some explanation of _specifically_ how this proposal achieves the benefits and use cases listed in the Motivation. It is currently not clear to me.

> We consider the implementation trustworthy after passing long comment period,

Author: ```suggestion
We consider the NIST candidates trustworthy after passing a long comment period,
```

considering that many competitors were eliminated by publishing weaknesses.

An alternative approach could be to provide low-level primitives from the reference library,
but that would decrease compatibility and security, since the reference implementation
will likely be supported by other software.

While lattice cryptography may require increased signature sizes, security-conscious
users will be willing to pay the price.

## Backward Compatibility

At the Plutus Core level, implementing this proposal induces no
backwards-incompatibility: the proposed new primitives do not break any existing
functionality or affect any other builtins. Likewise, at levels above Plutus
Core (such as `PlutusTx`), no existing functionality should be affected.

On-chain, this requires a hard fork.

## Path to Active

An implementation will require the following steps:

> 1. Linking to PQC library.

Reviewer: Which library? That is a big deal!

Author: ```suggestion
1. Linking to PQC library. We propose [PQClean](https://github.com/PQClean/PQClean) since it is a robust distribution of NIST-selected candidates. It also aims at cleanness and ease of formal verification.
```

Author: The reference C implementations of Post Quantum Crypto candidates from NIST are included in the library that facilitates their use.

2. Providing the patch implementing new Plutus primitives just like for [EcDSA and Schnorr](https://github.com/input-output-hk/plutus/pull/4368).
3. Providing the PQC primitives using `Crypto` interface for ease of use.

> ## Discussion full quantum resistance

Reviewer 2: Even then, it will be obvious that it was done illegally, and will be trivial to revert.

> ## Discussion full quantum resistance

Author: @iquerejeta Quantum and wallet scripts will need much more time to migrate.
We can do ledger and consensus migration as a last step, since they do not change API.

> ## Discussion full quantum resistance

Author: @L-as Concern about old generation of signatures is that Quantum Computing can hack them in the foreseeable future, so we need PQC wallets to protect against it.

Migration to PQC wallets will take some time, so it would be nice to provide PQC wallets even before entire blockchain is PQC-protected.

> This CIP _does not_ provide for full quantum resistance of the Cardano blockchain yet,

Reviewer: I wouldn't expect it to, it says it's about Plutus.

Author: @michaelpj Thanks Michael, but a lot of people (@iquerejeta, @rphair) asked about this more ambitious target.

in order to decrease cost of implementation.

SHA-256 hash is considered [_safe for a long-time yet_](https://cryptobook.nakov.com/quantum-safe-cryptography#quantum-safe-and-quantum-broken-crypto-algorithms), and any future upgrade would only need to consider SHA-384 as far as we know.
The main concern would be the public key signatures used for entire wallet infrastructure.
That is why we propose an _opt-in_ scheme, where users may gradually choose to migrate towards safer, post-quantum cryptography
over the course of few years.
That would make Cardano permit **quantum-resistant** smart contracts today at negligible cost,
and assure **long-term viability** of those smart contracts that will be deployed in the near future.

> [Long-term document storage](https://adastamp.co) is a killer application for the blockchain.

Author: Long-term document storage (notarization) needs a long term migration to new signature protocols.

This CIP proposes how to do this in general, beside handling post-quantum signatures as a special case.

> [Long-term document storage](https://adastamp.co) is a killer application for the blockchain.

Author: ```suggestion
[Long-term document storage](https://adastamp.co) is a killer application for the blockchain. It greatly benefits from standardized protocol of migration to better digital signature schemes.
```
